src/Mvc/Mvc.ApiExplorer/src/DefaultApiDescriptionProvider.cs

```diff
@@ -258,6 +262,20 @@ private void ProcessRouteParameters(ApiParameterContext context)
                         parameter.Source = BindingSource.Path;
                     }
                 }
+                else
+                {
+                    if (parameter.Source == BindingSource.Path &&
+                        parameter.ModelMetadata is DefaultModelMetadata defaultModelMetadata &&
+                        !defaultModelMetadata.Attributes.Attributes.OfType<IFromRouteMetadata>().Any())
+                    {
+                        // If we didn't see the parameter in the route and no FromRoute metadata is set, it probably means
+                        // the parameter binding source was inferred (InferParameterBindingInfoConvention)  
+                        // probably because another route to this action contains it as route parameter and
+                        // will be removed from the API description
+                        // https://github.com/dotnet/aspnetcore/issues/26234
+                        context.Results.RemoveAt(i);
```

Comment:
Does this situation only occur for parameters with default values? Given the scenario in the issue,

```C#
[HttpGet]
[Route("foo")]
[Route("bar/{id}")]
public IActionResult Get(int id = 0)
```

`id` could never come from the query string for the `foo` route, could it? Assuming not, this looks good.

Reply:
It does not because we stamp in `id`'s BindingSource as `FromRoute` due to this - https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Core/src/ApplicationModels/InferParameterBindingInfoConvention.cs#L109-L126 (essentially there's one ParameterInfo for all of the route templates, so we treat the templates as a union)

Reply:
Just to include an example: https://github.com/dotnet/aspnetcore/pull/39607#issuecomment-1017772958